### DISCUSSION TITLE
Why use manual tensor parallelism implementation instead of using something like deepspeed?

### DISCUSSION THREAD

**Discussion by vikigenius** [vllm-project/vllm#267] [Q&A]
I know it could have just been a design decision, but I would love to hear your rationale behind why you rolled your own implementation of model parallelism over using something like deepspeed as a dependency.

**Comment by zhuohan123** [comment]
Zero and FSDP communicate weights instead of activations. In inference, activations are small but weights are large. Therefore we choose tensor parallelism for efficiency.

**Reply by qingquansong** [reply]
For offline inference with prefill only case（such as embedding generation），this tp setting seems to be worse than deepspeed and fsdp.
